src/collect.jl

```diff
@@ -0,0 +1,34 @@
+collectcolumns(itr) = collectcolumns(itr, Base.iteratorsize(itr))
+
+function collectcolumns(itr, ::Union{Base.HasShape, Base.HasLength})
+    st = start(itr)
+    el, st = next(itr, st)
+    dest = similar(arrayof(typeof(el)), length(itr))
+    dest[1] = el
+    collect_to_columns!(dest, itr, 2, st)
+end
+
+function collect_to_columns!(dest::Columns{T, U}, itr, offs, st) where {T, U}
+    # collect to dest array, checking the type of each result. if a result does not
+    # match, widen the result type and re-dispatch.
+    i = offs
+    while !done(itr, st)
+        el, st = next(itr, st)
+        S = typeof(el)
+        if all((s <: t) for (s, t) in zip(S.parameters, T.parameters))
```

Comment:
Is this inferable if `eltype(itr)` is concrete or once `T` has become at least as wide as it? That's essential for performance in Base `map`.

Reply:
I don't have a strong enough intuition to understand what the compiler can optimize. Here Base uses `el isa T || typeof(el) === T`. What should I check exactly to test inferrability? Maybe making this a function `is_fieldwise_subtype(el::S,  ::Type{T}) = all((s <: t) for (s, t) in zip(S.parameters, T.parameters))` would help the compiler, I'm not sure

Reply:
`@code_warntype` will tell you whether the return type is inferred. Using a helper function could indeed help (maybe after marking it as `Base.@pure`, but the chances are that I'm wrong).

Reply:
I've checked with constant `eltype` and now it is tyoe stable. Somehow the way to achieve that was to use a generated function to check `is_fieldwise_subtype`, which we should do anyway for performance. It's not inferrable in the `NamedTuples` case (the initialization step `arrayof` which we already use in `map` is not inferrable), but I hope all the `NamedTuples` business will be much better in 0.7

Reply:
Still, should probably benchmark it to make sure it's doing OK compared with  the inference based `map`...

Reply:
There's probably a way to achieve the same result without a generated function, but these things are always tricky...